pkg/controller/nodemaintenance/nodemaintenance_controller_test.go

```diff
@@ -144,7 +152,245 @@ var _ = Describe("updateCondition", func() {
+			Expect(err).NotTo(HaveOccurred())
+			Expect(taintExist(taintedNode, "kubevirt.io/drain", corev1.TaintEffectNoSchedule)).To(Equal(true))
+			Expect(taintExist(taintedNode, "node.kubernetes.io/unschedulable", corev1.TaintEffectNoSchedule)).To(Equal(true))
+			Expect(taintExist(taintedNode, "test", corev1.TaintEffectPreferNoSchedule)).To(Equal(true))
+			Expect(len(taintedNode.Spec.Taints)).To(Equal(3))
+		})
+
+		It("should remove kubevirt NoSchedule taint and keep other existing taints", func() {
+			node, err := cs.CoreV1().Nodes().Get("node01", metav1.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(taintExist(node, "kubevirt.io/drain", corev1.TaintEffectNoSchedule)).To(Equal(false))
+			AddOrRemoveTaint(cs, node, true)
+			taintedNode, err := cs.CoreV1().Nodes().Get("node01", metav1.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(taintExist(taintedNode, "kubevirt.io/drain", corev1.TaintEffectNoSchedule)).To(Equal(true))
+			AddOrRemoveTaint(cs, taintedNode, false)
+			unTaintedNode, err := cs.CoreV1().Nodes().Get("node01", metav1.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(taintExist(unTaintedNode, "kubevirt.io/drain", corev1.TaintEffectNoSchedule)).To(Equal(false))
+			Expect(taintExist(unTaintedNode, "test", corev1.TaintEffectPreferNoSchedule)).To(Equal(true))
+			Expect(len(unTaintedNode.Spec.Taints)).To(Equal(1))
+		})
+	})
+
+	Context("Node maintenanace controller reconciles a maintenanace CR for a node in the cluster", func() {
+
+		It("should reconcile once without failing", func() {
+			reconcileMaintenance(nm)
+			checkSuccesfulReconcile()
+		})
+		It("should reconcile and cordon node", func() {
+			reconcileMaintenance(nm)
+			checkSuccesfulReconcile()
+			node, err := cs.CoreV1().Nodes().Get(nm.Spec.NodeName, metav1.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(node.Spec.Unschedulable).To(Equal(true))
+		})
+		It("should reconcile and taint node", func() {
+			reconcileMaintenance(nm)
+			checkSuccesfulReconcile()
+			node, err := cs.CoreV1().Nodes().Get(nm.Spec.NodeName, metav1.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(taintExist(node, "kubevirt.io/drain", corev1.TaintEffectNoSchedule)).To(Equal(true))
+		})
+		It("should fail on non existing node", func() {
+			nmCopy := nm.DeepCopy()
+			nmCopy.Spec.NodeName = "non-existing"
+			err := cl.Delete(context.TODO(), nm)
+			Expect(err).NotTo(HaveOccurred())
+			err = cl.Create(context.TODO(), nmCopy)
+			Expect(err).NotTo(HaveOccurred())
+			reconcileMaintenance(nmCopy)
+			checkFailedReconcile()
+		})
+
+	})
+})
+
+var _ = Describe("updateConditionWithLease", func() {
+
+	var r *ReconcileNodeMaintenance
+	var ctrl *gomock.Controller
+	var mockMaintenanceReconcile *MockReconcileHandler
+	var nm *kubevirtv1alpha1.NodeMaintenance
+	var cl client.Client
+	var cs *k8sfakeclient2.Clientset
+	var req reconcile.Request
+
+	setFakeClients := func() {
```

Comment:
Is this a pattern from somewhere?
Its a very long function to define inline - and use in other contexts.

Reply:
setFakeClients sets up the fake client, the fake client can return pre-recorded answers that are registered here in this function, so this function declares a very long array variable that is fed into the fake client. I copied the test setup  from regular unit tests, with the addition of the lease object. Is that the wrong approach?

Reply:
It shouldn't be inline as part of one test if its also used by other tests.